File: frontend/types/application.ts
// Mirrors the Industry-facing shape of `applications`
// (database/migrations/020_applications.sql) and
// backend/app/schemas/application.py. Keep all three in sync.
//
// Applicant identity: `profiles` RLS still gives Industry no direct read
// access to an applicant's `profiles` / `student_profiles` row — that is
// unchanged. `student_name` is resolved server-side (backend
// application_service, via the public.application_applicant_names RPC,
// database/migrations/036_application_applicant_names.sql) scoped to the
// same ownership predicate as the applications RLS policy itself, so it
// only ever carries `profiles.full_name` — no email, avatar, or other
// profile data. It can be null (lookup failure, or the student never set
// a full_name) — always fall back to `applicantRef`/`applicantDisplayName`
// rather than assuming it's present.

export const APPLICATION_STATUSES = [
  "APPLIED",
  "UNDER_REVIEW",
  "SHORTLISTED",
  "INTERVIEW_SCHEDULED",
  "SELECTED",
  "REJECTED",
  "WITHDRAWN",
] as const;
export type ApplicationStatus = (typeof APPLICATION_STATUSES)[number];

export const APPLICATION_STATUS_LABELS: Record<ApplicationStatus, string> = {
  APPLIED: "Applied",
  UNDER_REVIEW: "Under review",
  SHORTLISTED: "Shortlisted",
  INTERVIEW_SCHEDULED: "Interview scheduled",
  SELECTED: "Selected",
  REJECTED: "Rejected",
  WITHDRAWN: "Withdrawn",
};

export const OPPORTUNITY_TYPES = ["INTERNSHIP", "JOB"] as const;
export type OpportunityType = (typeof OPPORTUNITY_TYPES)[number];

export const OPPORTUNITY_TYPE_LABELS: Record<OpportunityType, string> = {
  INTERNSHIP: "Internship",
  JOB: "Job",
};

/** The statuses an Industry account can move an application to. Mirrors
 * backend `IndustrySettableStatus` — WITHDRAWN and APPLIED are excluded. */
export const INDUSTRY_SETTABLE_STATUSES = [
  "UNDER_REVIEW",
  "SHORTLISTED",
  "INTERVIEW_SCHEDULED",
  "SELECTED",
  "REJECTED",
] as const;
export type IndustrySettableStatus = (typeof INDUSTRY_SETTABLE_STATUSES)[number];

/** Valid Industry-driven transitions. Mirrors
 * application_service._STATUS_TRANSITIONS — used to only offer valid
 * actions in the UI (the backend still re-checks, so stale tabs get a
 * clean 409). */
export const APPLICATION_TRANSITIONS: Record<ApplicationStatus, IndustrySettableStatus[]> = {
  APPLIED: ["UNDER_REVIEW", "SHORTLISTED", "REJECTED"],
  UNDER_REVIEW: ["SHORTLISTED", "REJECTED"],
  SHORTLISTED: ["INTERVIEW_SCHEDULED", "REJECTED"],
  INTERVIEW_SCHEDULED: ["SELECTED", "REJECTED"],
  SELECTED: [],
  REJECTED: [],
  WITHDRAWN: [],
};

export const TRANSITION_LABELS: Record<IndustrySettableStatus, string> = {
  UNDER_REVIEW: "Move to review",
  SHORTLISTED: "Shortlist",
  INTERVIEW_SCHEDULED: "Schedule interview",
  SELECTED: "Mark selected",
  REJECTED: "Reject",
};

export interface ApplicationOpportunity {
  id: string;
  title: string;
  /** The posting's own lifecycle status (DRAFT/PUBLISHED/CLOSED/ARCHIVED). */
  status: string;
}

export interface Application {
  id: string;
  student_id: string;
  /** The applicant's display name (profiles.full_name), resolved
   * server-side. Null/undefined when resolution failed or the student has
   * no full_name — use `applicantDisplayName` rather than reading this
   * directly. */
  student_name?: string | null;
  industry_id: string;
  opportunity_type: OpportunityType;
  internship_id: string | null;
  job_id: string | null;
  status: ApplicationStatus;
  cover_note: string | null;
  /** Best-effort cache of the deterministic Skill Match score (Phase 9),
   * 0–100. Null until GET /api/v1/applications/{id}/match has run for it. */
  match_score: number | null;
  applied_at: string | null;
  created_at: string | null;
  updated_at: string | null;
  opportunity: ApplicationOpportunity | null;
}

/** GET /api/v1/applications/summary — drives the recruitment funnel. */
export interface ApplicationSummary {
  counts: Record<ApplicationStatus, number>;
  total: number;
}

/** The forward recruitment pipeline, in order — the funnel's stages. */
export const RECRUITMENT_PIPELINE: ApplicationStatus[] = [
  "APPLIED",
  "UNDER_REVIEW",
  "SHORTLISTED",
  "INTERVIEW_SCHEDULED",
  "SELECTED",
];

/** Statuses that leave the pipeline — shown separately from the funnel. */
export const RECRUITMENT_EXITS: ApplicationStatus[] = ["REJECTED", "WITHDRAWN"];

/** A short, privacy-safe label for an applicant, built only from the id —
 * the fallback used when no resolved name is available (RPC failure, or
 * the student never set a full_name). */
export function applicantRef(studentId: string): string {
  return `Applicant ${studentId.slice(0, 8)}`;
}

/** The applicant's real name when the backend resolved one, else the same
 * `applicantRef` truncated-id fallback every recruitment view already
 * used. Prefer this over `applicantRef` wherever an application's
 * applicant is displayed. */
export function applicantDisplayName(application: {
  student_id: string;
  student_name?: string | null;
}): string {
  const name = application.student_name?.trim();
  return name ? name : applicantRef(application.student_id);
}

// ===========================================================================
// Skill Match (Phase 9) — GET /api/v1/applications/{id}/match
// Deterministic and advisory. No LLM. The backend is the sole source of the
// score; the frontend never recomputes it. Mirrors
// backend/app/schemas/application.py ApplicationMatchResponse / MatchSkill.
// ===========================================================================

export const MATCH_SKILL_STATUSES = ["MATCHED", "NEEDS_IMPROVEMENT", "MISSING"] as const;
export type MatchSkillStatus = (typeof MATCH_SKILL_STATUSES)[number];

export const MATCH_SKILL_STATUS_LABELS: Record<MatchSkillStatus, string> = {
  MATCHED: "Matched",
  NEEDS_IMPROVEMENT: "Needs improvement",
  MISSING: "Missing",
};

export const MATCH_RECOMMENDATIONS = ["STRONG", "GOOD", "PARTIAL", "LOW"] as const;
export type MatchRecommendation = (typeof MATCH_RECOMMENDATIONS)[number];

export const MATCH_RECOMMENDATION_LABELS: Record<MatchRecommendation, string> = {
  STRONG: "Strong match",
  GOOD: "Good match",
  PARTIAL: "Partial match",
  LOW: "Low match",
};

export interface MatchSkill {
  skill_id: string;
  skill_name: string;
  required_level: string;
  importance: string;
  candidate_has: boolean;
  candidate_level: string | null;
  candidate_verified: boolean;
  status: MatchSkillStatus;
}

export interface ApplicationMatch {
  application_id: string;
  score: number;
  recommendation: MatchRecommendation;
  skill_coverage: string;
  required_count: number;
  matched_count: number;
  needs_improvement_count: number;
  missing_count: number;
  matched_skills: MatchSkill[];
  needs_improvement_skills: MatchSkill[];
  missing_skills: MatchSkill[];
}

File: backend/app/schemas/application.py
"""Pydantic schemas for the Industry side of applications
(database/migrations/020_applications.sql -- the unified internship + job
application table).

An application row is created by a STUDENT applying to a published
internship/job. Industry's involvement is read + a single mutable field:
`status`. Everything else on the row is immutable to Industry, enforced
by the `prevent_application_identity_change` trigger and by only ever
sending `{"status": ...}` from the service.

Student identity: `profiles` RLS still only permits a user to read their
own row (001_profiles.sql) -- that is unchanged. `student_name` is
resolved server-side through `public.application_applicant_names`
(036_application_applicant_names.sql), a SECURITY DEFINER function scoped
to the exact same "Industry can view applications to their own postings"
predicate as the applications table's own RLS SELECT policy, so it can
never name a student for an application the caller doesn't already own.
It returns only `profiles.full_name` -- no email, avatar, or any other
profile data. When name resolution fails (or the student has no
full_name), `student_name` is None and the frontend falls back to a
truncated `student_id` reference.
"""

from typing import Literal

from pydantic import BaseModel, ConfigDict

# database/migrations/020_applications.sql -- applications.status CHECK
ApplicationStatus = Literal[
    "APPLIED",
    "UNDER_REVIEW",
    "SHORTLISTED",
    "INTERVIEW_SCHEDULED",
    "SELECTED",
    "REJECTED",
    "WITHDRAWN",
]
OpportunityType = Literal["INTERNSHIP", "JOB"]

APPLICATION_STATUSES: tuple[str, ...] = (
    "APPLIED",
    "UNDER_REVIEW",
    "SHORTLISTED",
    "INTERVIEW_SCHEDULED",
    "SELECTED",
    "REJECTED",
    "WITHDRAWN",
)

# The status values an INDUSTRY account is allowed to set. WITHDRAWN is
# deliberately absent -- that transition belongs to the student, and the
# `prevent_student_status_override` trigger is the mirror of this rule on
# the database side. APPLIED is absent -- it is the initial state only.
IndustrySettableStatus = Literal[
    "UNDER_REVIEW",
    "SHORTLISTED",
    "INTERVIEW_SCHEDULED",
    "SELECTED",
    "REJECTED",
]


class ApplicationOpportunity(BaseModel):
    """The internship or job an application was submitted to, resolved
    through the caller's own postings (Industry owns these rows, so RLS
    permits the join). `status` here is the *posting's* lifecycle status
    (DRAFT/PUBLISHED/CLOSED/ARCHIVED), not the application's."""

    id: str
    title: str
    status: str


class ApplicationResponse(BaseModel):
    id: str
    student_id: str
    # Resolved via public.application_applicant_names -- see module
    # docstring. None if resolution fails or the student has no full_name.
    student_name: str | None = None
    industry_id: str
    opportunity_type: str
    internship_id: str | None = None
    job_id: str | None = None
    status: str
    cover_note: str | None = None
    # Populated later by the AI matching service (Phase 8) -- not written here.
    match_score: float | None = None
    applied_at: str | None = None
    created_at: str | None = None
    updated_at: str | None = None
    opportunity: ApplicationOpportunity | None = None


class ApplicationListResponse(BaseModel):
    applications: list[ApplicationResponse]


class ApplicationSummaryResponse(BaseModel):
    """Per-status counts of the caller's own applications, for the
    recruitment funnel. `counts` carries every status (0 when none)."""

    counts: dict[str, int]
    total: int


SkillMatchStatus = Literal["MATCHED", "NEEDS_IMPROVEMENT", "MISSING"]
MatchRecommendation = Literal["STRONG", "GOOD", "PARTIAL", "LOW"]


class MatchSkill(BaseModel):
    """One required skill of the posting, with how the applicant covers it.
    Only skills the caller's own posting requires ever appear here."""

    skill_id: str
    skill_name: str
    required_level: str
    importance: str
    candidate_has: bool
    candidate_level: str | None = None
    candidate_verified: bool = False
    status: SkillMatchStatus


class ApplicationMatchResponse(BaseModel):
    """Advisory only. A deterministic, reproducible skill-fit summary for
    one application. Never connected to status transitions. No LLM, no
    stored explanation -- the structured result is authoritative."""

    application_id: str
    score: int  # 0-100
    recommendation: MatchRecommendation
    skill_coverage: str  # e.g. "8 / 10"
    required_count: int
    matched_count: int
    needs_improvement_count: int
    missing_count: int
    matched_skills: list[MatchSkill]
    needs_improvement_skills: list[MatchSkill]
    missing_skills: list[MatchSkill]


class ApplicationStatusUpdate(BaseModel):
    """PATCH /api/v1/applications/{id}/status body.

    `status` is a restricted Literal, so an out-of-range value, WITHDRAWN,
    APPLIED, or a smuggled `student_id` / `industry_id` / `internship_id`
    all fail validation with a 422 before any handler runs
    (`extra="forbid"`).
    """

    model_config = ConfigDict(extra="forbid")

    status: IndustrySettableStatus
